# Benchmark Summary Statistics
---

Description:

    This notebook performs the summary statisticks results for all benchmark functions (one at a time).
    The optimization process is repeated N=50 times (with default settings) and then we collect the results.
    
---

### Import python libraries and set up the project directory

In [8]:
import os, sys
import numpy as np

PROJECT_DIR = os.path.abspath('..')
sys.path.append(PROJECT_DIR)

### Import the custom PSO code

In [9]:
from star_pso.population.swarm import Swarm
from star_pso.population.particle import Particle
from star_pso.engines.standard_pso import StandardPSO
from star_pso.utils.auxiliary import cost_function

### Load all the Benchmark problems

In [10]:
# Dimensions: 1D
from star_pso.benchmarks.equal_maxima import EqualMaxima
from star_pso.benchmarks.five_uneven_peak_trap import FiveUnevenPeakTrap
from star_pso.benchmarks.uneven_decreasing_maxima import UnevenDecreasingMaxima

# Dimensions: 2D
from star_pso.benchmarks.himmelblau import Himmelblau
from star_pso.benchmarks.six_hump_camel_back import SixHumpCamelBack

# Dimensions: N >= 2D
from star_pso.benchmarks.shubert import Shubert
from star_pso.benchmarks.vincent import Vincent
from star_pso.benchmarks.rastrigin import Rastrigin

In [11]:
# Select one test problem.
benchmark = Shubert(n_dim=3)

### Define the cost function

Note: All the optimization problems are solved as maximizations.

In [12]:
@cost_function(minimize=False)
def func_benchmark(x_array: np.ndarray, **kwargs) -> float:
    # Return the solution.
    return benchmark.func(x_array).item()
# _end_def_

### Setup the experiment

In [13]:
# Number of repeatitions.
n_repeats = 50

# Swarm population size.
POP_SIZE = 1600

# Maximum function evaluations.
MAX_F_EVALS = 400_000

# Store the results.
results = n_repeats * [0]

# Display Info.
print(f"Testing function: {benchmark.name}\n")

# Go time ...
for i in range(n_repeats):

    # Display current iteration.
    print(f"Running optimization: {i+1} out of {n_repeats} ... ", flush=True)

    # Get an initial (new) sample.
    x_sample = benchmark.sample_random_positions(n_pos=POP_SIZE)

    # Initial swarm population.
    swarm_t0 = Swarm([Particle(x) for x in x_sample])
    
    # Create the StandardPSO object that will carry on the optimization.
    test_PSO = StandardPSO(initial_swarm = swarm_t0, obj_func = func_benchmark,
                           x_min = benchmark.x_min, x_max = benchmark.x_max)
    # Run the optimization.
    test_PSO.run(max_it = 800, f_max_eval = MAX_F_EVALS, verbose = False,
                 options = {"w0": 0.75, "c1": 1.50, "c2": 1.50, "mode": "multimodal"})
    
    # Obtain the number of the global optima that were found for epsilon = 1.0e-4.
    gopt_found, gopt_total = benchmark.search_for_optima(test_PSO.swarm.best_n(POP_SIZE))

    # Update the results.
    results[i] = gopt_found

    # Blank line to declatter the screen.
    print(" ")
# _end_for_

print("Done.")

Testing function: Shubert_2D

Running optimization: 1 out of 50 ... 


11/28/2025 22:01:34 INFO: Initial f_optimal = 108.4742
11/28/2025 22:01:55 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 21.769 seconds.
 
Running optimization: 2 out of 50 ... 


11/28/2025 22:01:55 INFO: Initial f_optimal = 155.8324
11/28/2025 22:02:11 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.138 seconds.
 
Running optimization: 3 out of 50 ... 


11/28/2025 22:02:11 INFO: Initial f_optimal = 180.4484
11/28/2025 22:02:25 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.770 seconds.
 
Running optimization: 4 out of 50 ... 


11/28/2025 22:02:25 INFO: Initial f_optimal = 157.7049
11/28/2025 22:02:41 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.205 seconds.
 
Running optimization: 5 out of 50 ... 


11/28/2025 22:02:41 INFO: Initial f_optimal = 160.2146
11/28/2025 22:02:57 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.676 seconds.
 
Running optimization: 6 out of 50 ... 


11/28/2025 22:02:57 INFO: Initial f_optimal = 179.8236
11/28/2025 22:03:12 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.738 seconds.
 
Running optimization: 7 out of 50 ... 


11/28/2025 22:03:12 INFO: Initial f_optimal = 168.1109
11/28/2025 22:03:27 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.730 seconds.
 
Running optimization: 8 out of 50 ... 


11/28/2025 22:03:27 INFO: Initial f_optimal = 155.5454
11/28/2025 22:03:42 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.819 seconds.
 
Running optimization: 9 out of 50 ... 


11/28/2025 22:03:42 INFO: Initial f_optimal = 131.0793
11/28/2025 22:03:56 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.673 seconds.
 
Running optimization: 10 out of 50 ... 


11/28/2025 22:03:57 INFO: Initial f_optimal = 113.8735
11/28/2025 22:04:11 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.987 seconds.
 
Running optimization: 11 out of 50 ... 


11/28/2025 22:04:12 INFO: Initial f_optimal = 99.8164
11/28/2025 22:04:26 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.945 seconds.
 
Running optimization: 12 out of 50 ... 


11/28/2025 22:04:27 INFO: Initial f_optimal = 163.8756
11/28/2025 22:04:41 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 14.979 seconds.
 
Running optimization: 13 out of 50 ... 


11/28/2025 22:04:42 INFO: Initial f_optimal = 183.7012
11/28/2025 22:04:58 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.596 seconds.
 
Running optimization: 14 out of 50 ... 


11/28/2025 22:04:58 INFO: Initial f_optimal = 150.8492
11/28/2025 22:05:13 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.014 seconds.
 
Running optimization: 15 out of 50 ... 


11/28/2025 22:05:13 INFO: Initial f_optimal = 185.7165
11/28/2025 22:05:28 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.067 seconds.
 
Running optimization: 16 out of 50 ... 


11/28/2025 22:05:28 INFO: Initial f_optimal = 174.5085
11/28/2025 22:05:44 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.681 seconds.
 
Running optimization: 17 out of 50 ... 


11/28/2025 22:05:44 INFO: Initial f_optimal = 154.4269
11/28/2025 22:05:59 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.308 seconds.
 
Running optimization: 18 out of 50 ... 


11/28/2025 22:05:59 INFO: Initial f_optimal = 168.5261
11/28/2025 22:06:15 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 15.427 seconds.
 
Running optimization: 19 out of 50 ... 


11/28/2025 22:06:15 INFO: Initial f_optimal = 183.2758
11/28/2025 22:06:32 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.911 seconds.
 
Running optimization: 20 out of 50 ... 


11/28/2025 22:06:32 INFO: Initial f_optimal = 183.8085
11/28/2025 22:06:51 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.775 seconds.
 
Running optimization: 21 out of 50 ... 


11/28/2025 22:06:51 INFO: Initial f_optimal = 186.7276
11/28/2025 22:07:07 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.878 seconds.
 
Running optimization: 22 out of 50 ... 


11/28/2025 22:07:08 INFO: Initial f_optimal = 186.1663
11/28/2025 22:07:24 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.153 seconds.
 
Running optimization: 23 out of 50 ... 


11/28/2025 22:07:24 INFO: Initial f_optimal = 175.7019
11/28/2025 22:07:40 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.281 seconds.
 
Running optimization: 24 out of 50 ... 


11/28/2025 22:07:40 INFO: Initial f_optimal = 97.3472
11/28/2025 22:07:56 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.328 seconds.
 
Running optimization: 25 out of 50 ... 


11/28/2025 22:07:56 INFO: Initial f_optimal = 144.0114
11/28/2025 22:08:13 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.379 seconds.
 
Running optimization: 26 out of 50 ... 


11/28/2025 22:08:13 INFO: Initial f_optimal = 175.9870
11/28/2025 22:08:29 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.731 seconds.
 
Running optimization: 27 out of 50 ... 


11/28/2025 22:08:30 INFO: Initial f_optimal = 180.1034
11/28/2025 22:08:46 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.575 seconds.
 
Running optimization: 28 out of 50 ... 


11/28/2025 22:08:46 INFO: Initial f_optimal = 173.9445
11/28/2025 22:09:04 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 17.936 seconds.
 
Running optimization: 29 out of 50 ... 


11/28/2025 22:09:04 INFO: Initial f_optimal = 122.5988
11/28/2025 22:09:21 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.878 seconds.
 
Running optimization: 30 out of 50 ... 


11/28/2025 22:09:21 INFO: Initial f_optimal = 183.8172
11/28/2025 22:09:38 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.873 seconds.
 
Running optimization: 31 out of 50 ... 


11/28/2025 22:09:38 INFO: Initial f_optimal = 145.1100
11/28/2025 22:09:55 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.841 seconds.
 
Running optimization: 32 out of 50 ... 


11/28/2025 22:09:55 INFO: Initial f_optimal = 134.0701
11/28/2025 22:10:12 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.785 seconds.
 
Running optimization: 33 out of 50 ... 


11/28/2025 22:10:12 INFO: Initial f_optimal = 134.5501
11/28/2025 22:10:28 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.851 seconds.
 
Running optimization: 34 out of 50 ... 


11/28/2025 22:10:29 INFO: Initial f_optimal = 183.2246
11/28/2025 22:10:47 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.134 seconds.
 
Running optimization: 35 out of 50 ... 


11/28/2025 22:10:47 INFO: Initial f_optimal = 169.7252
11/28/2025 22:11:04 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.842 seconds.
 
Running optimization: 36 out of 50 ... 


11/28/2025 22:11:04 INFO: Initial f_optimal = 135.6413
11/28/2025 22:11:21 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 17.060 seconds.
 
Running optimization: 37 out of 50 ... 


11/28/2025 22:11:21 INFO: Initial f_optimal = 154.8597
11/28/2025 22:11:38 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.963 seconds.
 
Running optimization: 38 out of 50 ... 


11/28/2025 22:11:38 INFO: Initial f_optimal = 171.5524
11/28/2025 22:11:56 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.780 seconds.
 
Running optimization: 39 out of 50 ... 


11/28/2025 22:11:57 INFO: Initial f_optimal = 162.4205
11/28/2025 22:12:15 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.741 seconds.
 
Running optimization: 40 out of 50 ... 


11/28/2025 22:12:15 INFO: Initial f_optimal = 171.5089
11/28/2025 22:12:34 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.801 seconds.
 
Running optimization: 41 out of 50 ... 


11/28/2025 22:12:34 INFO: Initial f_optimal = 183.5036
11/28/2025 22:12:58 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 24.074 seconds.
 
Running optimization: 42 out of 50 ... 


11/28/2025 22:12:58 INFO: Initial f_optimal = 141.8642
11/28/2025 22:13:16 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 17.840 seconds.
 
Running optimization: 43 out of 50 ... 


11/28/2025 22:13:16 INFO: Initial f_optimal = 180.0982
11/28/2025 22:13:33 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.865 seconds.
 
Running optimization: 44 out of 50 ... 


11/28/2025 22:13:33 INFO: Initial f_optimal = 180.4017
11/28/2025 22:13:50 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.889 seconds.
 
Running optimization: 45 out of 50 ... 


11/28/2025 22:13:50 INFO: Initial f_optimal = 121.5186
11/28/2025 22:14:07 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 17.239 seconds.
 
Running optimization: 46 out of 50 ... 


11/28/2025 22:14:07 INFO: Initial f_optimal = 179.8934
11/28/2025 22:14:25 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 17.337 seconds.
 
Running optimization: 47 out of 50 ... 


11/28/2025 22:14:25 INFO: Initial f_optimal = 184.0242
11/28/2025 22:14:41 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.850 seconds.
 
Running optimization: 48 out of 50 ... 


11/28/2025 22:14:41 INFO: Initial f_optimal = 184.2156
11/28/2025 22:15:00 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 19.040 seconds.
 
Running optimization: 49 out of 50 ... 


11/28/2025 22:15:01 INFO: Initial f_optimal = 170.2036
11/28/2025 22:15:17 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 16.863 seconds.
 
Running optimization: 50 out of 50 ... 


11/28/2025 22:15:17 INFO: Initial f_optimal = 161.5035
11/28/2025 22:15:35 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 333


Final f_optimal = 186.7309
run: elapsed time = 18.031 seconds.
 
Done.


In [14]:
# Compute the "Peak Ratio".
peak_ratio = sum(results) / (gopt_total * n_repeats)

print(f"PR = {peak_ratio}")

PR = 0.9833333333333333


In [15]:
# Compute the "Success Rate".
success_rate = sum([r == gopt_total for r in results]) / n_repeats

print(f"SR = {success_rate}")

SR = 0.74


### Optimization Results

Accuracy level: $\epsilon = 1.0^{-4}$

| Benchmark Function | Dimensions | No. Global opt. | MaxFEs  | Population Size | Parameters | PR | SR |
|:---|:---|:---|:---|:---|:---|:---|:---|
| Equal Maxima | 1 | 5 | 50_000| 120 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Five Uneven Peak Trap | 1 | 2 | 50_000 | 120 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Uneven Decreasing Maxima | 1 | 1 | 50_000 | 85 | (0.70, 1.50, 1.50) | 1.0 | 1.0 |
| Himmelblau | 2 | 4 | 50_000 | 100 | (0.70, 1.50, 1.50) |1.0 | 1.0 |
| Six Hump Camel Back | 2 | 2 | 50_000 | 85 | (0.70, 1.50, 1.50) |1.0 | 1.0 |
| Rastrigin | 2 / 4| 6 / 12 | 100_000 / 200_000 | 150 / 400 | (0.70, 1.50, 1.50) |1.000 / 0.998 | 1.000 / 0.980 |
| Shubert | 2 / 3 | 18 / 81 | 200_000 / 400_000 | 600 / 1600 | (0.75, 1.50, 1.50) |0.983 / 0.637 | 0.740 / 0.000 |
| Vincent | 2 / 3 | 36 / 216 | 200_000 / 400_000 | 2000 / 4000 | (0.70, 1.50, 1.50) |0.744 / 0.392 | 0.000 / 0.000 |

### End of file